language/extensions/move-table-extension/sources/Table.move

```diff
@@ -0,0 +1,44 @@
+/// Type of large-scale storage tables.
+module Extensions::Table {
+    struct Table<phantom K, phantom V> has store {
+        handle: u128
+    }
+
+    /// Create a new Table.
+    public fun new<K, V>(): Table<K, V> {
+        Table{handle: new_table_handle()}
+    }
+
+    native fun new_table_handle(): u128;
+
+    /// Destroy a table. The table must be empty to succeed.
+    native fun destroy(table: Table);
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    native public fun borrow<K, V>(table: &Table<K, V>, key: &K): &V;
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    native public fun borrow_mut<K, V>(table: &mut Table<K, V>, key: &K): &mut V;
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Insert the pair (`key`, `default`) first if there is no entry for `key`.
+    public fun borrow_mut_with_default<K: copy + drop, V: drop>(table: &mut Table<K, V>, key: K, default: V): &mut V {
+        if (!contains(table, &key)) {
+            insert(table, copy key, default)
+        };
+        borrow_mut(table, &key)
+    }
+
+    /// Remove from `table` and return the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    native public fun remove<K, V>(table: &mut Table<K, V>, key: &K): (K, V);
+
+    /// Returns true iff `table` contains an entry for `key`.
+    native public fun contains<K, V>(table: &Table<K, V>, key: &K): bool;
+
+    /// Insert the pair (`key`, `val`) to `table`.
+    /// Aborts if there is already an entry for `key`.
+    native public fun insert<K, V>(table: &mut Table<K, V>, key: K, val: V);
```

Comment:
Yeah, this is the right design for insert. Many libraries return back the value if it exists. I'm smart contract development, I found that there's likely different behavior if there are two values at the same key, e.g. merge. If we return the value in move, we then need to deal with dropping option none or other nonsense in the common case, which isn't ergonomic.